eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ledger.hs

```diff
@@ -396,16 +396,26 @@ ledgerTransition = do
               , certState
               , StrictSeq.fromStrict $ txBody ^. certsTxBodyL
               )
-        let wdrlAddrs = Map.keysSet . unWithdrawals $ tx ^. bodyTxL . withdrawalsTxBodyL
-            wdrlCreds = Set.map raCredential wdrlAddrs
-            dUnified = dsUnified $ certDState certStateAfterCERTS
-            delegatedAddrs = DRepUView dUnified
-
-        -- TODO: Finish this implementation once we are in bootstrap phase:
-        -- https://github.com/IntersectMBO/cardano-ledger/issues/4092
-        when False $ do
-          all (`UMap.member` delegatedAddrs) wdrlCreds
-            ?! ConwayWdrlNotDelegatedToDRep (wdrlCreds Set.\\ Map.keysSet (dRepMap dUnified))
+
+        let checkWdrlsKeyHashesDelegated = do
+              let
+                isKeyHash = \case
+                  KeyHashObj _ -> True
+                  _ -> False
+                wdrlsKeyHashes =
+                  Set.filter isKeyHash $
+                    Set.map raCredential $
```

Comment:
We can add a more helpful and reusable function `credKeyHash` to `Cardano.Ledger.Credential` instead:

```haskell
credKeyHash :: Credential r c -> Maybe (KeyHash r c)
credKeyHash = \case
  KeyHashObj hk -> Just hk
  ScriptHashObj _ -> Nothing
```

```suggestion
                wdrlsKeyHashes =
                    Map.mapMaybe (credKeyHash . raCredential) $
```

Then we can reuse it for implementing `credKeyHashWitness`:
```haskell
credKeyHashWitness :: Credential r c -> Maybe (KeyHash 'Witness c)
credKeyHashWitness = fmap asWitness . credKeyHash
```

Adjust further here for the DRep delegation check as well.

Reply:
Nice idea!  Or even 
```
credKeyHashWitness = credKeyHash . asWitness
```